### ISSUE TITLE
americanas.com.br immediately sends RST_STREAM

### ISSUE THREAD

**Issue by batata004** [curl/curl#13546]
### I did this

The website americanas.com.br is the largest ecommerce in brazil after amazon.com. For some reason, simply requesting the main page returns with error. It's not a protection or any security measure of the website, curl is buggy. Even using http1.1 flag does not work.

You just need to execute the code below in your SSH:

curl 'https://www.americanas.com.br/' \
  -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7' \
  -H 'accept-language: pt-BR,pt;q=0.9,en-US;q=0.8,en;q=0.7' \
  -H 'cache-control: no-cache' \
  -H 'pragma: no-cache' \
  -H 'sec-ch-ua: "Google Chrome";v="123", "Not:A-Brand";v="8", "Chromium";v="123"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  -H 'sec-fetch-dest: document' \
  -H 'sec-fetch-mode: navigate' \
  -H 'sec-fetch-site: same-origin' \
  -H 'sec-fetch-user: ?1' \
  -H 'upgrade-insecure-requests: 1' \
  -H 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36'

### I expected the following

I expected the request to be completed. However, I get this error;

curl: (92) HTTP/2 stream 0 was not closed cleanly: INTERNAL_ERROR (err 2)

I think this website might be doing something not according to RFC in the headers, however it's the second largest ecommerce in brazil, curl could allow some flag to bypass RFC compliance.

### curl/libcurl version

latest

### operating system

centos stream

**Comment by bagder** [comment]
> It's not a protection or any security measure of the website

How can you tell the reason behind their odd behavior?

> curl is buggy.

That's not necessarily a correct observation here. We need much more data before we can tell.

It rather looks like the website is "buggy". When requested with HTTP/2 it sends an immediate stream reset and when asked to do HTTP/1.1 it does not respond.

TLS fingerprinting perhaps?

**Comment by icing** [comment]
The sites sends me a real response when adding `-H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:125.0) Gecko/20100101 Firefox/125.0'` to the command line.

**Comment by batata004** [comment]
@bagder it's not curl fingerprinting, I tried with `curl impersonate` and the bug still happens. Why did you add the tag `not-a-curl-bug` when it clearly is a curl bug? If it's not a bug, even if the server is doing something wrong with the headers, I think you could add a flag so the user can disable RFC compliance. This is the second largest ecommerce in Brazil and `curl` does not work with it.


@icing I used `-H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:125.0) Gecko/20100101 Firefox/125.0'` as you said, and I still got the error response. Here is my SSH command with the header you suggested. I executed the command below on my SSH on CentOS stream and also with PHP and I got the same error.

curl 'https://www.americanas.com.br/' \
  -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7' \
  -H 'accept-language: pt-BR,pt;q=0.9,en-US;q=0.8,en;q=0.7' \
  -H 'cache-control: no-cache' \
  -H 'pragma: no-cache' \
  -H 'sec-ch-ua: "Google Chrome";v="123", "Not:A-Brand";v="8", "Chromium";v="123"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  -H 'sec-fetch-dest: document' \
  -H 'sec-fetch-mode: navigate' \
  -H 'sec-fetch-site: same-origin' \
  -H 'sec-fetch-user: ?1' \
  -H 'upgrade-insecure-requests: 1' \
  -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:125.0) Gecko/20100101 Firefox/125.0'

**Comment by bagder** [comment]
This is not a curl bug because the server side is clearly actively deciding to close the connection based on user-agent sniffing.

Like for @icing, it works fine for me with both 8.7.1 and current master when I do:

~~~
curl -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:125.0) Gecko/20100101 Firefox/125.0' 'https://www.americanas.com.br/'

**Comment by batata004** [comment]
Hi,

I have an updated Centos Stream and the latest version is not 8.x but 7.x. My `dnf -y update` shows everything is updated. Do you have any idea why Centos Stream is not updating itself to the latest?

Here is my curl version:

[root@localhost ~]# curl --version
curl 7.76.1 (x86_64-redhat-linux-gnu) libcurl/7.76.1 OpenSSL/3.2.1 zlib/1.2.11 b                      rotli/1.0.9 libidn2/2.3.0 libpsl/0.21.1 (+libidn2/2.3.0) libssh/0.10.4/openssl/z                      lib nghttp2/1.43.0
Release-Date: 2021-04-14
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mq                      tt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos L                      argefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets

**Comment by jay** [comment]
some distros like centos maintain an older version of libcurl (and possibly curl) and it's required by the system. i don't recommend overwriting it. if you want the latest version you will have to look for an alternative package repo or build it yourself. i would install the latest version to a different prefix  like /usr/local which is safe from affecting the system packages

**Comment by batata004** [comment]
@jay That's pretty sad. Indeed, I checked online and it's not safe to change the default `curl` version that came with the OS. I will try to install it in another directory, as you said. Thank you.

**Comment by bagder** [comment]
@batata004 So when you said latest, you meant curl 7.76.1? A three years old version...

**Comment by batata004** [comment]
@bagder To me, the latest is the latest my OS provides me. If `curl` maintainers dont care about pushing the latest into the OSes they support, it's not me to blame. I think `curl` maintainers should push Centos to provide the latest to all users. What's the purpose of you fixing multiple bugs and security holes if you dont spend time to make it available to the broader audience? I mean, I am talking about CentOS, I am not talking about a weird linux distro nobody uses on their servers.

**Comment by dfandrich** [comment]
You should educate yourself about how Linux distributions, and Open Source in
general, work before coming here to criticize those who provide you absolutely
free software.

**Comment by batata004** [comment]
Of course.

**Comment by jay** [comment]
no evidence this is a curl bug, closing.

**Comment by dsommers** [comment]
@batata004 I haven't dug up the CentOS Stream packaging policy .... but here is the one related to package updates for Fedora EPEL packages: https://docs.fedoraproject.org/en-US/epel/epel-policy/#_digest

> The goal is to have packages in EPEL that enhances the Enterprise Linux distributions the packages were build against without disturbing or replacing packages from that distribution. The packages in the repository should, if possible, be maintained in similar ways to the Enterprise Packages they were built against. **In other words: have a mostly stable set of packages that normally do not change at all and only changes if there are good reasons for it — so no "hey, there is a new version, it builds, let’s ship it" mentality.**

Each distribution has their own packaging policy which all package maintainers must adhere to.

**Also notice this crucially important detail:** Upstream project developers (like curl devs) are most typically not much involved in distribution packaging of their project.  All distributions typically have their own community of package maintainers owning that responsibility.  In most cases, it's even different persons for each distribution.  Often the upstream developers don't even know all the distributions a project is packaged for, as there is no requirement for package maintainers to be involved at all in the upstream project.

Bottom line: If there are issues in the curl version in your CentOS Stream distribution ... you need to reach out to the CentOS Stream package maintainer and discuss your issue there.  Especially when the upstream project cannot reproduce the issues you have.

**Comment by pemensik** [comment]
@batata004 your OS also provides you its own issue tracker, which is not this one. You should report issue with distribution you are using to its issue tracker. I have failed to find similar issue at [issues.redhat.com](https://issues.redhat.com/issues/?jql=component%3Dcurl%20AND%20text%20~%20americanas.com.br). Not for RHEL or CentOS Stream. If you even need to try latest version, we have [Fedora Rawhide](https://src.fedoraproject.org/rpms/curl) for it. Including [podman/docker images](https://quay.io/repository/fedora/fedora?tab=tags).  CentOS is not distribution shipping latest releases and that is not going to change.

But I am quite sure the result would be the same as here. That is not an issue to fix in curl, in any version.

**Comment by bagder** [comment]
> If curl maintainers want to stick with the RFC, I think you could at least provide a flag for those people who dont care about RFC, people who just want to get the response back of their requests without further problems or discussions of what is ideal, what is correct... All the largest browsers dont care about most RFC, they pass the response back to the users even if the server does some weird thing (as long as it does not pose a security issue, like a misconfigured SSL handshack and so on).

@batata004 the server is actively trying to detect what client that is used and have decided that if it thinks its not a browser it kills the transfer. That has nothing to do with following RFCs or not. curl does not fail because of a "minor and simple server misconfiguration".

I also doubt that you know what "All the largest browsers" do or don't do when it comes to HTTP/2 protocol spec compliance. I have worked hands-on with one browser HTTP/2 implementation and talked HTTP/2 details with developers of other browsers, and your description does not match my view.

I think the best approach is if you make a pull-request yourself and show us exactly what kind of edits you are talking about. After all, code is often better than words.

The nonsense in this issue stops now.
